test/distributed/fsdp/test_fsdp_clip_grad_norm.py

```diff
@@ -39,67 +41,133 @@
         )
-        total_norms_after_clip_local = _collect_total_grad_norm_local(
-            local_model, norm_type
+        ddp_model = DDP(local_model, device_ids=[self.rank])
+        fsdp_kwargs = {
+            "auto_wrap_policy": functools.partial(
+                transformer_auto_wrap_policy,
+                transformer_layer_cls={
+                    TransformerEncoderLayer,
+                    TransformerDecoderLayer,
+                },
+            ),
+            "cpu_offload": CPUOffload(offload_params=offload_params),
+            "use_orig_params": use_orig_params,
+        }
+        fsdp_model = TransformerWithSharedParams.init(
+            self.process_group,
+            FSDPInitMode.RECURSIVE,
+            CUDAInitMode.CUDA_BEFORE,
+            deterministic=True,
+            fsdp_kwargs=fsdp_kwargs,
         )
-        self.assertTrue(total_norms_after_clip_fsdp <= norm_cap)
-        self.assertEqual(total_norms_after_clip_local, total_norms_after_clip_fsdp)
+        LR = 1e-2
+        ddp_optim = torch.optim.Adam(ddp_model.parameters(), lr=LR)
+        fsdp_optim = torch.optim.Adam(fsdp_model.parameters(), lr=LR)
+        device = torch.device("cuda")
+        LARGE_FACTOR = 100
+        inp = ddp_model.module.get_input(device)
+        for model in (ddp_model, fsdp_model):
+            out = model(*inp)
+            loss = model.module.get_loss(inp, out)
+            loss.backward()
 
-    @skip_if_lt_x_gpu(2)
-    @parametrize("norm_type", [2.0, inf])
-    @parametrize("nested_fsdp", [True, False])
-    @parametrize(
-        "cpu_offload",
-        [CPUOffload(offload_params=True), CPUOffload(offload_params=False)],
-    )
-    def test_fsdp_clip_grad_norm(self, norm_type, nested_fsdp, cpu_offload):
-        """Test FSDP with clip grad norm."""
-        self._run_fsdp_one_iteration(norm_type, nested_fsdp, cpu_offload)
+        # Multiply gradients by a large factor to ensure that gradients will
+        # actually be clipped
+        for param in itertools.chain(ddp_model.parameters(), fsdp_model.parameters()):
+            if param.grad is not None:  # gradients may be `None` for `use_orig_params=True`
+                param.grad *= LARGE_FACTOR
+        orig_ddp_grads = [param.grad.detach().clone() for param in ddp_model.parameters()]
+        orig_fsdp_grads = [
+            param.grad.detach().clone() if param.grad is not None else None
+            for param in fsdp_model.parameters()
+        ]
+
+        ddp_total_norm = torch.nn.utils.clip_grad_norm_(
+            ddp_model.parameters(), max_norm=max_norm, norm_type=norm_type,
+        )
+        fsdp_total_norm = fsdp_model.clip_grad_norm_(max_norm=max_norm, norm_type=norm_type)
+        self.assertEqual(ddp_total_norm, fsdp_total_norm)
 
+        # Check that the gradients were modified by `clip_grad_norm_()`
+        for param, orig_grad in zip(ddp_model.parameters(), orig_ddp_grads):
+            assert not torch.equal(param.grad, orig_grad)
```

Comment:
nit: instead of `assert`, use `self.assertNotEqual`?

Reply:
I use `assert` when it is part of the test infrastructure itself. If this `assert` fails, then the unit test is meaningless; it does not indicate that the unit test failed.